src/urbit.ts

```diff
@@ -0,0 +1,468 @@
+      console.log(error);
+    });
+  }
+  
+
+  /**
+   * Returns (and initializes, if necessary) the SSE pipe for the appropriate channel.
+   */
+  eventSource(): EventSource {
+    if (!this.sseClient) {
+      const sseOptions = {
+        headers: {}
+      };
+      if (isBrowser) {
+        sseOptions.withCredentials = true;
+      } else if (isNode) {
+        sseOptions.headers.Cookie = this.cookie;
+      }
+      this.sseClient = new EventSource(this.channelUrl, sseOptions);
+      this.sseClient?.addEventListener('message', (event: MessageEvent) => {
+        if (this.verbose) {
+          console.log('Received SSE: ', event);
+        }
+        this.ack(Number(event.lastEventId));
+        if (event.data && JSON.parse(event.data)) {
+          const data: any = JSON.parse(event.data);
+          if (data.response === 'poke' && this.outstandingPokes.has(data.id)) {
+            const funcs = this.outstandingPokes.get(data.id);
+            if (data.hasOwnProperty('ok')) {
+              funcs.success();
+            } else if (data.hasOwnProperty('err')) {
+              funcs.fail(data.err);
+            } else {
+              console.error('Invalid poke response', data);
+            }
+            this.outstandingPokes.delete(data.id);
+          } else if (data.response === 'subscribe' || 
+            (data.response === 'poke' && this.outstandingSubscriptions.has(data.id))) {
+            const funcs = this.outstandingSubscriptions.get(data.id);
+            if (data.hasOwnProperty('err')) {
+              funcs.err(data.err);
+              this.outstandingSubscriptions.delete(data.id);
+            }
+          } else if (data.response === 'diff' && this.outstandingSubscriptions.has(data.id)) {
+            const funcs = this.outstandingSubscriptions.get(data.id);
+            funcs.event(data.json);
+          } else if (data.response === 'quit' && this.outstandingSubscriptions.has(data.id)) {
+            const funcs = this.outstandingSubscriptions.get(data.id);
+            funcs.quit(data);
+            this.outstandingSubscriptions.delete(data.id);
+          } else {
+            console.log('Unrecognized response', data);
+          }
+          // An incoming message, for example:
+          // {
+          //   id: 10,
+          //   json: {
+          //     'chat-update' : { // This is where we hook our "on" handlers like "onChatUpdate"
+          //       message: {
+          //         envelope: {
+          //           author: 'zod',
+          //           letter: {
+          //             text: 'hi'
+          //           },
+          //           number: 10,
+          //           uid: 'saludhafhsdf',
+          //           when: 124459
+          //         },
+          //         path: '/~zod/mailbox'
+          //       }
+          //     }
+          //   }
+          // }
+        }
+      });
+      this.sseClient?.addEventListener('error', function(event: Event) {
```

Comment:
(i'm getting really picky I'm sorry!) It's probably better style to use `this.sseClient!.addEventListener` the zap will assert to TS that `this.sseClient` is non-null, whilst not actually changing the semantics of the property access.

Reply:
Nice I've wondered about that